### PR TITLE
Block Bindings: Add block `uses_context` defined by block bindings sources

### PR DESCRIPTION
**Pull request by SantosGuillamot**
This pull request adds some logic to add the necessary `uses_context` defined by block bindings sources. Each source defines the context it needs and it is added to the blocks that are using the block bindings API.

Apart from that, it adapts the post meta and pattern overrides sources for this change.

Finally, it removes the `pattern/overrides` from the `block.json` because it is handled this way now.

### Testing instructions

0. Register a custom field in your site to test it. You can use a code snippet like this:
```
register_meta(
	'post',
	'url_custom_field',
	array(
		'show_in_rest' => true,
		'single'       => true,
		'type'         => 'string',
		'default'	   => 'https://wpmovies.dev/wp-content/uploads/2023/04/goncharov-poster-original-1-682x1024.jpeg',
	)
);
```

**Post meta source test**
1. Go to a page and insert an image block.
2. Go to the Code editor and connect the image to the custom field by adding the metadata bindings property. It should look something like this:
```
<!-- wp:image {"id":134,"sizeSlug":"large","linkDestination":"none","metadata":{"bindings":{"url":{"source":"core/post-meta","args":{"key":"url_custom_field"}},"alt":{"source":"core/post-meta","args":{"key":"text_custom_field"}}}}} -->
<figure class="wp-block-image size-large"><img src="https://wpmovies.dev/wp-content/uploads/2023/04/goncharov-poster-original-1-682x1024.jpeg" alt="Alt content" class="wp-image-134"/></figure>
<!-- /wp:image -->
```
3. Save the page and check that in the frontend, the URL defined in the custom field is used.
4. Repeat the process for paragraph, heading, and button blocks.

**Pattern overrides test**
1. Go to Site Editor -> Patterns -> New pattern.
2. Create a synced pattern and give it a name.
3. Add a paragraph with a default value and, in the advanced panel, enable the "Allow instance overrides" setting.
4. Repeat the process with a heading, an image, and a button.
5. Go to a page and insert the pattern created.
6. Modify the value of the blocks and save the page.
7. Go to the frontend and check that the modified values are used.

---

Trac ticket: https://core.trac.wordpress.org/ticket/60525

---
**This Pull Request is for code review only. Please keep all other discussion in the Trac ticket. Do not merge this Pull Request. See [GitHub Pull Requests for Code Review](https://make.wordpress.org/core/handbook/contribute/git/github-pull-requests-for-code-review/) in the Core Handbook for more details.**

### CODE REVIEW

**Comment by SantosGuillamot** [comment]
I guess, according to the tests, that I shouldn't modify the `block.json` properties, right? Is that imported from Gutenberg somehow?

**Comment by gziolo** [comment]
> I guess, according to the tests, that I shouldn't modify the block.json properties, right? Is that imported from Gutenberg somehow?

That needs to happen in the Gutenberg plugin. It isn't as important, so that can be synced later during the 6.5 Beta phase as code quality improvement. You also need to plan to remove these `useContext` entries from the Gutenberg plugin by syncing back the PHP changes proposed in this patch.

File: src/wp-includes/block-bindings/post-meta.php [hunk 1]
```diff
@@ -51,6 +48,7 @@ function _register_block_bindings_post_meta_source() {
 		array(
 			'label'              => _x( 'Post Meta', 'block bindings source' ),
 			'get_value_callback' => '_block_bindings_post_meta_get_value',
+			'uses_context'       => array( 'postId', 'postType' ),
```
Comment (gziolo) [inline src/wp-includes/block-bindings/post-meta.php:52]: Is `postType` ever used on the client? I don't see it being referenced on the server.

File: src/wp-includes/class-wp-block-bindings-registry.php [hunk 2]
```diff
@@ -144,6 +157,25 @@ public function register( string $source_name, array $source_properties ) {
...
+			'register_block_type_args',
+			function ( $args, $block_name ) use ( $source ) {
+				$allowed_blocks = $this->allowed_blocks;
+
+				if ( empty( $allowed_blocks[ $block_name ] ) || empty( $source->uses_context ) ) {
+					return $args;
+				}
+				$original_use_context = isset( $args['uses_context'] ) ? $args['uses_context'] : array();
+				// Use array_values to reset the array keys.
+				$args['uses_context'] = array_values( array_unique( array_merge( $original_use_context, $source->uses_context ) ) );
```
Comment (gziolo) [inline src/wp-includes/class-wp-block-bindings-registry.php]: The comment included makes me feel that it needs to be covered with unit tests to ensure it works correctly.

File: src/wp-includes/class-wp-block.php [hunk 3]
```diff
@@ -235,14 +235,7 @@ private function process_block_bindings() {
...
 
-		// Allowed blocks that support block bindings.
-		// TODO: Look for a mechanism to opt-in for this. Maybe adding a property to block attributes?
-		$allowed_blocks = array(
-			'core/paragraph' => array( 'content' ),
-			'core/heading'   => array( 'content' ),
-			'core/image'     => array( 'url', 'title', 'alt' ),
-			'core/button'    => array( 'url', 'text', 'linkTarget', 'rel' ),
-		);
+		$allowed_blocks = WP_Block_Bindings_Registry::get_instance()->allowed_blocks;
```
Comment (gziolo) [inline src/wp-includes/class-wp-block.php]: That part leaks the temporary list of allowed blocks so it would become a public API. We need to ensure it doesn't happen and find an alternative approach.

File: src/wp-includes/block-bindings/post-meta.php (hunk 1 above)
Comment (SantosGuillamot) [reply src/wp-includes/block-bindings/post-meta.php:52]: Yes, it is used to retrieve the custom field depending on the `postType`: [link](https://github.com/WordPress/gutenberg/blob/06cfe8c95015189c762afd73acd31c3c5c76ee17/packages/editor/src/bindings/post-meta.js#L22-L27).

File: src/wp-includes/class-wp-block-bindings-registry.php (hunk 2 above)
Comment (SantosGuillamot) [reply src/wp-includes/class-wp-block-bindings-registry.php]: Makes sense. I'll work on that 🙂

File: src/wp-includes/class-wp-block.php (hunk 3 above)
Comment (SantosGuillamot) [reply src/wp-includes/class-wp-block.php]: You're right. We can go back to the hardcoded variable.

File: src/wp-includes/class-wp-block-bindings-registry.php [hunk 4]
```diff
@@ -32,6 +32,19 @@ final class WP_Block_Bindings_Registry {
 	 */
 	private static $instance = null;
 
+	/**
+	 * Allowed block that can use the block bindings API.
```
Comment (michalczaplinski) [inline src/wp-includes/class-wp-block-bindings-registry.php]: Nitpick: block**s** (plural)
```suggestion
	 * Allowed blocks that can use the block bindings API.
```

File: src/wp-includes/class-wp-block-bindings-registry.php [hunk 5]
```diff
@@ -32,6 +32,19 @@ final class WP_Block_Bindings_Registry {
...
+	 *
+	 * @since 6.5.0
+	 * @var WP_Block_Bindings_Registry[]
```
Comment (michalczaplinski) [inline src/wp-includes/class-wp-block-bindings-registry.php]: Nitpick: This should just be `array`.

File: src/wp-includes/class-wp-block-bindings-source.php [hunk 6]
```diff
@@ -60,6 +68,9 @@ public function __construct( string $name, array $source_properties ) {
 		$this->name               = $name;
 		$this->label              = $source_properties['label'];
 		$this->get_value_callback = $source_properties['get_value_callback'];
+		if ( isset( $source_properties['uses_context'] ) ) {
```
Comment (michalczaplinski) [inline src/wp-includes/class-wp-block-bindings-source.php]: I think that we should move this `isset()` check next to the [other checks](https://github.com/SantosGuillamot/wordpress-develop/blob/c4fb33fdaee7388c6246b3857d10696666f620ab/src/wp-includes/class-wp-block-bindings-registry.php#L123-L141)
in `WP_Block_Bindings_Registry`.

Comment (michalczaplinski) [reply src/wp-includes/class-wp-block-bindings-source.php]: And let's cover it with a unit test as well 
🙂. (Should be almost a copy-paste of the
existing ones)

**Comment by michalczaplinski** [comment]
Let's also make sure that we update the docstrings in:
- `register()` method - [link](https://github.com/SantosGuillamot/wordpress-develop/blob/c4fb33fdaee7388c6246b3857d10696666f620ab/src/wp-includes/class-wp-block-bindings-registry.php#L66)
- `register_block_bindings_source()` function - [link](https://github.com/SantosGuillamot/wordpress-develop/blob/c4fb33fdaee7388c6246b3857d10696666f620ab/src/wp-includes/block-bindings.php#L75)

File: src/wp-includes/class-wp-block-bindings-source.php (hunk 6 above)
Comment (SantosGuillamot) [reply src/wp-includes/class-wp-block-bindings-source.php]: My idea is to make the `uses_context` optional because some sources won't need to use it. Do you think it shouldn't be that case? It would make sense to add a check that, if it exists, it is an array though.

I'll work on unit tests as well 🙂

Comment (gziolo) [reply src/wp-includes/class-wp-block-bindings-source.php]: Yes, it should be optional and default to `null`.

File: src/wp-includes/class-wp-block-bindings-registry.php (hunk 5 above)
Comment (gziolo) [reply src/wp-includes/class-wp-block-bindings-registry.php]: I think you need only the list of blocks here:

```php
private $supported_blocks = array(
	'core/paragraph',
	'core/heading',
	'core/image',
	'core/button',
);
```

In the future, we will have to figure out a way for a block to declare if it supports block bindings. It could be opt in or out depending how the API evolves.

File: src/wp-includes/class-wp-block-bindings-source.php (hunk 6 above)
Comment (michalczaplinski) [reply src/wp-includes/class-wp-block-bindings-source.php]: Right, that makes sense! It should be optional indeed 🙂

File: src/wp-includes/class-wp-block-bindings-registry.php (hunk 2 above)
Comment (gziolo) [reply src/wp-includes/class-wp-block-bindings-registry.php]: We might not need to call `array_values`. Every place in the code I checked leaves the array keys unmodified. What was the reason you added `array_values` call?

Comment (SantosGuillamot) [reply src/wp-includes/class-wp-block-bindings-registry.php]: I will check it again, but without `array_values` it was returning some errors caused by the array keys. If I remember correctly, when we merge two arrays like these:

```
$original_uses_context = array( 'postId' );
$source_uses_context = array( 'postId', 'postType' );
$merged_array = array_unique( array_merge( $original_use_context, source_uses_context ) );
```
The result array has an issue with the keys:
```
[0] => 'postId',
[2] => 'postType',
```
It seems to skip the repeated elements (`postId`), so I was using `array_values` to reset that.

Anyway, I will take another look at it.

File: src/wp-includes/class-wp-block-bindings-registry.php (hunk 5 above)
Comment (SantosGuillamot) [reply src/wp-includes/class-wp-block-bindings-registry.php]: This should be fixed now: [link](https://github.com/WordPress/wordpress-develop/pull/6079/files#diff-3fd1b8d9e55d4fe09b5f48002d360d3ba9d338d6f33ef3959828424a590b4947R35-R46).

File: src/wp-includes/class-wp-block.php (hunk 3 above)
Comment (SantosGuillamot) [reply src/wp-includes/class-wp-block.php]: I went back to the hardcoded variable: [link](https://github.com/WordPress/wordpress-develop/pull/6079/files#diff-c569b1ecb11a007a563788a481710b51e0cb21cd1d1885490bf9c91777842f23R238-R243).

**Comment by SantosGuillamot** [comment]
> Let's also make sure that we update the docstrings in:

I made these changes to the files you shared. Thanks for pointing that out 🙂 

I've been discussing the implementation with @gziolo, and we've modified it a bit to include a new `get_block_type_uses_context` hook instead of using `register_block_type_args`. This should be more reliable, especially when external plugins create their own sources.

Although I'm still reviewing how to make it work properly.

File: src/wp-includes/class-wp-block-bindings-source.php (hunk 6 above)
Comment (SantosGuillamot) [reply src/wp-includes/class-wp-block-bindings-source.php]: I've added the check [here](https://github.com/WordPress/wordpress-develop/pull/6079/files#diff-3fd1b8d9e55d4fe09b5f48002d360d3ba9d338d6f33ef3959828424a590b4947R154-R162).

I'll check the unit tests.

File: src/wp-includes/class-wp-block-bindings-registry.php (hunk 2 above)
Comment (SantosGuillamot) [reply src/wp-includes/class-wp-block-bindings-registry.php]: By the way, it causes an error in the editor, not the server.

**Comment by SantosGuillamot** [comment]
I've tested the new approach registering a new source like this in an external plugin (without Gutenberg plugin) and everything seems to work as expected 🙂 

```php
add_action('init', function() {
	register_block_bindings_source('plugin/test', [
		'label' => 'Plugin Test',
		'get_value_callback' => function(array $args, $block_instance, $name) {
			return 'Post Type: ' . $block_instance->context['postType'] . ' Post Id: ' . $block_instance->context['postId'];
		},
		'uses_context' => ['postType', 'postId'],
	]);
});
```

File: src/wp-includes/class-wp-block-bindings-registry.php [hunk 7]
```diff
@@ -151,6 +151,16 @@ public function register( string $source_name, array $source_properties ) {
 			return false;
 		}
 
+		/* Validate that the uses_context parameter is an array */
```
Comment (gziolo) [inline src/wp-includes/class-wp-block-bindings-registry.php]: Nit, coding style:
```suggestion
		// Validate that the uses_context parameter is an array.
```

File: src/wp-includes/block-bindings/post-meta.php [hunk 8]
```diff
@@ -17,18 +17,15 @@
  *                           Example: array( "key" => "foo" ).
  * @return mixed The value computed for the source.
```
Comment (gziolo) [inline src/wp-includes/block-bindings/post-meta.php:19]: Nit: the param description is missing. You can copy from the other source.

File: src/wp-includes/class-wp-block-bindings-registry.php (hunk 7 above)
Comment (SantosGuillamot) [reply src/wp-includes/class-wp-block-bindings-registry.php]: I copied and pasted from the other comments. I guess it applies there as well, right?

**Comment by gziolo** [comment]
> I've tested the new approach registering a new source like this in an external plugin (without Gutenberg plugin) and everything seems to work as expected 🙂

Can you also run a quick check with the Gutenberg plugin when activated? It looks like the similar changes for variations introduced with https://core.trac.wordpress.org/changeset/57315 caused some compatibility issues in the Gutenberg plugin, as discussed in https://core.trac.wordpress.org/ticket/59969.

@kt-12 or @joemcgill, we are looking at replicating the same pattern for `uses_context` in `WP_Block_Type` as you landed for `variations. The reason is that we want to be able dynamically extend `uses_context` based on another registry for the block binding sources. We can't guarantee that all block types get registered after plugins finish registering all sources. I think, on principle, the problem is very similar to defining block variations by extenders. Do you have any concerns with the proposed solution?

File: src/wp-includes/class-wp-block-bindings-registry.php (hunk 7 above)
Comment (SantosGuillamot) [reply src/wp-includes/class-wp-block-bindings-registry.php]: I just made these changes.

File: src/wp-includes/block-bindings/post-meta.php (hunk 8 above)
Comment (SantosGuillamot) [reply src/wp-includes/block-bindings/post-meta.php:19]: I've added it [here](https://github.com/WordPress/wordpress-develop/pull/6079/files).

File: src/wp-includes/class-wp-block-bindings-registry.php (hunk 2 above)
Comment (kt-12) [reply src/wp-includes/class-wp-block-bindings-registry.php]: @SantosGuillamot  
You can try the following, a much cleaner code with the same effect.

```php
$original_uses_context = array( 'postId' );
$source_uses_context = array( 'postId', 'postType' );
$merged_array = array_unique( $original_uses_context + $source_uses_context );
```

**Comment by SantosGuillamot** [comment]
> Can you also run a quick check with the Gutenberg plugin when activated?

I've been running more tests registering a `plugin/testing` new source and it seems to work in these scenarios:
* Register it in Gutenberg and use WP 6.5.
* Register it in Gutenberg and use WP 6.4.3.
* Register it in an external plugin and use WP 6.5 without Gutenberg.
* Register it in an external plugin and use WP 6.5 with Gutenberg.
* Register it in an external plugin and use WP 6.4.3 with Gutenberg.

Any other use case I'm missing?

For Gutenberg I'm using the latest version of [this pull request](https://github.com/WordPress/gutenberg/pull/58554).

File: src/wp-includes/class-wp-block-type.php [hunk 9]
```diff
@@ -398,6 +402,10 @@ public function __isset( $name ) {
 			return true;
 		}
 
+		if ( 'uses_context' === $name ) {
```
Comment (kt-12) [inline src/wp-includes/class-wp-block-type.php]: nit-pick: I feel we should avoid adding separate blocks for look-alike conditions.

Instead, club both `variations` and `uses_context` checks in one condition using `in_array`

File: src/wp-includes/class-wp-block-type.php [hunk 10]
```diff
@@ -422,6 +430,11 @@ public function __set( $name, $value ) {
 			return;
 		}
 
+		if ( 'uses_context' === $name ) {
```
Comment (kt-12) [inline src/wp-includes/class-wp-block-type.php]: Now that I look more closely, I believe we can get rid of both `variations` and `uses_context` conditions here. It will be handled by the next block for non-deprecated properties.

Comment (SantosGuillamot) [reply src/wp-includes/class-wp-block-type.php]: I removed it in [this commit](https://github.com/WordPress/wordpress-develop/pull/6079/commits/0c4a795247b372b5efb103350ae532dc9058d584). Let me know if that's not what you suggested.

File: src/wp-includes/class-wp-block-type.php (hunk 9 above)
Comment (SantosGuillamot) [reply src/wp-includes/class-wp-block-type.php]: Sure! I just did it in [this commit](https://github.com/WordPress/wordpress-develop/pull/6079/commits/f4af4f7f794f47822b2335f779b790404317d1bd).

File: src/wp-includes/class-wp-block-bindings-registry.php (hunk 2 above)
Comment (SantosGuillamot) [reply src/wp-includes/class-wp-block-bindings-registry.php]: I've tested using `$original_uses_context + `source_uses_context`, but it doesn't work as expected.

Apart from that, I added a test where I am checking if the index is correct: [link](https://github.com/SantosGuillamot/wordpress-develop/blob/f2b3b3af8491de34530fb2802029c867a7406f93/tests/phpunit/tests/block-bindings/wpBlockBindingsRegistry.php#L357-L390).

**Comment by SantosGuillamot** [comment]
I've created a bunch of tests to check `uses_context`:
* Test that the value defined in `uses_context` is available for the source: [link](https://github.com/SantosGuillamot/wordpress-develop/blob/f2b3b3af8491de34530fb2802029c867a7406f93/tests/phpunit/tests/block-bindings/render.php#L124-L158).
* Test that using a string in `uses_context` fails: [link](https://github.com/SantosGuillamot/wordpress-develop/blob/f2b3b3af8491de34530fb2802029c867a7406f93/tests/phpunit/tests/block-bindings/wpBlockBindingsRegistry.php#L175-L181).
* Test that the result of registering a source has the correct values: [link](https://github.com/SantosGuillamot/wordpress-develop/blob/f2b3b3af8491de34530fb2802029c867a7406f93/tests/phpunit/tests/block-bindings/wpBlockBindingsRegistry.php#L200-L206).
* Test that multiple sources, with shared context, merge it properly: [link](https://github.com/SantosGuillamot/wordpress-develop/blob/f2b3b3af8491de34530fb2802029c867a7406f93/tests/phpunit/tests/block-bindings/wpBlockBindingsRegistry.php#L357-L390).

File: src/wp-includes/class-wp-block-bindings-source.php (hunk 6 above)
Comment (SantosGuillamot) [reply src/wp-includes/class-wp-block-bindings-source.php]: As explained here, I added a few test to ensure it works as expected: [link](https://github.com/WordPress/wordpress-develop/pull/6079#issuecomment-1945789396).

Comment (gziolo) [reply src/wp-includes/class-wp-block-bindings-source.php]: There are two options in the constructor:

- keep adding `isset` checks for the settings that are optional like at the moment
- ensure that we pass only allowed settings in `$source_properties` in the registry and add `foreach` loop similar to:
  https://github.com/WordPress/wordpress-develop/blob/7b24083859c37d8ac75b9549f458bc389610eaef/src/wp-includes/class-wp-block-type.php#L576C3-L578C4

File: tests/phpunit/tests/block-bindings/render.php [hunk 11]
```diff
@@ -114,6 +114,49 @@ public function test_passing_arguments_to_source() {
...
+
+		$this->assertSame(
+			'Value: source context value',
+			$block->attributes['content'],
+			'Value: source context value'
+		);
+		$this->assertSame(
+			'<p>Value: source context value</p>',
+			trim( $result ),
+			'Value: source context value'
```
Comment (gziolo) [inline tests/phpunit/tests/block-bindings/render.php]: The 3rd param should explain what the assertion does so it's easier to locate which one failed. It should be different than the other one.

File: tests/phpunit/tests/block-bindings/wpBlockBindingsRegistry.php [hunk 12]
```diff
@@ -321,4 +346,46 @@ public function test_is_registered_for_known_source() {
...
+		);
+
+		$block_registry          = WP_Block_Type_Registry::get_instance();
+		$paragraph_block_context = $block_registry->get_registered( 'core/paragraph' )->uses_context;
+		// Check that the resulting `uses_context` contains the values from both sources.
+		$this->assertContains( 'commonContext', $paragraph_block_context );
+		$this->assertContains( 'sourceOneContext', $paragraph_block_context );
+		$this->assertContains( 'sourceTwoContext', $paragraph_block_context );
+		// Check that the resulting `uses_context` has unique values.
+		$this->assertCount( count( array_unique( $paragraph_block_context ) ), $paragraph_block_context );
```
Comment (gziolo) [inline tests/phpunit/tests/block-bindings/wpBlockBindingsRegistry.php]: It would be better to use a custom block that has a predictable `uses_context` (does not change over time) so we could hardcode the expected count.

File: tests/phpunit/tests/block-bindings/render.php [hunk 13]
```diff
@@ -114,6 +114,49 @@ public function test_passing_arguments_to_source() {
...
+				'uses_context'       => array( 'sourceContext' ),
+			)
+		);
+
+		$block_content = <<<HTML
+		<!-- wp:paragraph {"metadata":{"bindings":{"content":{"source":"test/source", "args": {"key": "test"}}}}} -->
+		<p>This should not appear</p>
+		<!-- /wp:paragraph -->
+		HTML;
+		$parsed_blocks = parse_blocks( $block_content );
```
Comment (gziolo) [inline tests/phpunit/tests/block-bindings/render.php:144]: Something is wrong here and needs to be investigated further:

Screenshot 2024-02-15 at 11 52 25

Comment (gziolo) [reply tests/phpunit/tests/block-bindings/render.php:144]: It fails only on older versions of PHP, so my guess is that in line 143 you need to remove spaces:

```php
HTML;
```

File: tests/phpunit/tests/block-bindings/render.php (hunk 11 above)
Comment (SantosGuillamot) [reply tests/phpunit/tests/block-bindings/render.php]: You're totally right. I just changed it.

File: tests/phpunit/tests/block-bindings/wpBlockBindingsRegistry.php (hunk 12 above)
Comment (gziolo) [reply tests/phpunit/tests/block-bindings/wpBlockBindingsRegistry.php]: Ok, we need to use one of the supported blocks. Still, I would put a number rather than compute it on the fly, which makes the test harder to follow.

Comment (SantosGuillamot) [reply tests/phpunit/tests/block-bindings/wpBlockBindingsRegistry.php]: I thought about that, but the problem is that the list of supported blocks, where the source `uses_context` is added, is hardcoded right now. So I don't know how to do it.

Comment (gziolo) [reply tests/phpunit/tests/block-bindings/wpBlockBindingsRegistry.php]: Another option is to run `$this->assertCount()` before registering every source to show how the number grows.

File: src/wp-includes/class-wp-block-bindings-source.php (hunk 6 above)
Comment (SantosGuillamot) [reply src/wp-includes/class-wp-block-bindings-source.php]: If I understood you correctly, I believe I prefer the second option. I made [this commit](https://github.com/WordPress/wordpress-develop/pull/6079/commits/07a7317882354bb68304a34a6c929c12b45d424f) to illustrate how I understood it. We can always revert it.

_PD: I included unnecessary code I was using for testing, I just removed it in other commit_ 😅

File: tests/phpunit/tests/block-bindings/render.php (hunk 13 above)
Comment (SantosGuillamot) [reply tests/phpunit/tests/block-bindings/render.php:144]: Just did it in [this commit](https://github.com/WordPress/wordpress-develop/pull/6079/commits/edaaffad0cceeef98ca9d982189057a25a364710). Let's see if that solves the problem. I guess it should because it is used in other tests that way.

File: tests/phpunit/tests/block-bindings/wpBlockBindingsRegistry.php (hunk 12 above)
Comment (SantosGuillamot) [reply tests/phpunit/tests/block-bindings/wpBlockBindingsRegistry.php]: I've hardcoded the expected increment between the original context and the new one in [this commit](https://github.com/WordPress/wordpress-develop/pull/6079/commits/03e66ee954843b1e9e12710fab0a8f93cba95c1b). Let me know if you feel that is easier to understand.

File: tests/phpunit/tests/block-bindings/wpBlockBindingsRegistry.php [hunk 14]
```diff
@@ -321,4 +346,48 @@ public function test_is_registered_for_known_source() {
...
+		);
+
+		$new_uses_context = $block_registry->get_registered( 'core/paragraph' )->uses_context;
+		// Check that the resulting `uses_context` contains the values from both sources.
+		$this->assertContains( 'commonContext', $new_uses_context );
+		$this->assertContains( 'sourceOneContext', $new_uses_context );
+		$this->assertContains( 'sourceTwoContext', $new_uses_context );
+		// Check that the resulting `uses_context` only added 3 items: commonContext, sourceOneContext, and sourceTwoContext.
+		$this->assertSame( count( $original_uses_context ) + 3, count( $new_uses_context ) );
+		// Check that the index is correct and it doesn't skip numbers after merging. Needed for the editor.
```
Comment (gziolo) [inline tests/phpunit/tests/block-bindings/wpBlockBindingsRegistry.php]: Nit: how about the following comments?
```suggestion
		// Checks that the resulting `uses_context` contains the values from both sources.
		$this->assertContains( 'commonContext', $new_uses_context );
		$this->assertContains( 'sourceOneContext', $new_uses_context );
		$this->assertContains( 'sourceTwoContext', $new_uses_context );
		// Checks that the resulting `uses_context` added 3 unique items.
		$this->assertSame( count( $original_uses_context ) + 3, count( $new_uses_context ) );
		// Checks that the array isn't sparse to prevent issues in the editor.
```

I used 3rd person as in other comments.

File: src/wp-includes/class-wp-block-bindings-registry.php (hunk 2 above)
Comment (gziolo) [reply src/wp-includes/class-wp-block-bindings-registry.php]: Let's leave it as is. We have an unit test in place that ensures that the editor doesn't crash so we can refactor that in the future safely.

File: src/wp-includes/class-wp-block-bindings-source.php (hunk 6 above)
Comment (gziolo) [reply src/wp-includes/class-wp-block-bindings-source.php]: Yes, it's exactly how it could work 👍🏻 

It would be nice to add a unit test for the registry that asserts that it errors when an unknown property is provided.

---

At the same time, we are closing the source object for an extension, which I'm a bit unsure about, given we might need to change during beta if we figure out it prevents iterations in the Gutenberg plugin. We also miss the following annotation on the class if we decide to go that path:

https://github.com/WordPress/wordpress-develop/blob/7b24083859c37d8ac75b9549f458bc389610eaef/src/wp-includes/class-wp-block-type.php#L17

File: tests/phpunit/tests/block-bindings/wpBlockBindingsRegistry.php [hunk 15]
```diff
@@ -321,4 +346,48 @@ public function test_is_registered_for_known_source() {
 		$result = $this->registry->is_registered( self::$test_source_name );
 		$this->assertTrue( $result );
 	}
+
+	/**
+	 * Test merging `uses_context` from multiple sources.
+	 *
+	 * @ticket 60525
+	 *
+	 * @covers ::register_block_bindings_source
```
Comment (gziolo) [inline tests/phpunit/tests/block-bindings/wpBlockBindingsRegistry.php:355]:
```suggestion
	 * @covers ::register_block_bindings_source
	 * @covers WP_Block_Type::get_uses_context
```

File: src/wp-includes/class-wp-block-bindings-registry.php [hunk 16]
```diff
@@ -159,7 +220,7 @@ public function unregister( string $source_name ) {
 		if ( ! $this->is_registered( $source_name ) ) {
 			_doing_it_wrong(
 				__METHOD__,
-				/* translators: %s: Block bindings source name. */
+				// translators: %s: Block bindings source name.
```
Comment (gziolo) [inline src/wp-includes/class-wp-block-bindings-registry.php]: Can we revert the changes applied to the hints provided to translators? It looks like they all use the `/* */` syntax for some reasons I don't fully understand.

File: src/wp-includes/class-wp-block-bindings-registry.php [hunk 17]
```diff
@@ -137,13 +163,48 @@ public function register( string $source_name, array $source_properties ) {
...
+		add_filter(
+			'get_block_type_uses_context',
+			function ( $uses_context, $block_type ) use ( $source ) {
+				if ( ! in_array( $block_type->name, $this->supported_blocks, true ) || empty( $source->uses_context ) ) {
+					return $uses_context;
+				}
+				// Use array_values to reset the array keys.
+				$merged_uses_context = array_values( array_unique( array_merge( $uses_context, $source->uses_context ) ) );
+
+				return $merged_uses_context;
```
Comment (gziolo) [inline src/wp-includes/class-wp-block-bindings-registry.php]: Nit, the local variable doesn't seem to add much value:
```suggestion
				return array_values( array_unique( array_merge( $uses_context, $source->uses_context ) ) );
```

**Review by gziolo — APPROVED** [review]
I left some additional feedback to consider, but it's already in good shape overall. Thank you for adding all the changes and covering the unit tests.

File: src/wp-includes/class-wp-block-bindings-registry.php [hunk 18]
```diff
@@ -137,13 +163,48 @@ public function register( string $source_name, array $source_properties ) {
...
 			$source_name,
 			$source_properties
 		);
 
 		$this->sources[ $source_name ] = $source;
 
+		// Add `uses_context` defined by block bindings sources.
+		add_filter(
+			'get_block_type_uses_context',
+			function ( $uses_context, $block_type ) use ( $source ) {
```
Comment (gziolo) [inline src/wp-includes/class-wp-block-bindings-registry.php:195]: One more thing here and not a blocker. I believe we should use a class method and unregister the filter when unregistering the source. It would also help with documenting, as we could put all details from inline comments in the PHPDoc.

I'm not entirely sure how that would look in practice, but happy to explore together options. It can be another patch.

File: src/wp-includes/class-wp-block-bindings-source.php (hunk 6 above)
Comment (SantosGuillamot) [reply src/wp-includes/class-wp-block-bindings-source.php]: I am unsure about restricting it as well. I can expect we might want to increase it in Gutenberg after external developers create their own sources the same way we need now to include `uses_context`.

Comment (SantosGuillamot) [reply src/wp-includes/class-wp-block-bindings-source.php]: On the other hand, if we allow source to use their own properties, could we face backwards-compatibility issues in the future?

Imagine a source is using a custom property `format`. And in the future, we want to include it as part of the core properties and we use it somehow as we are doing with `uses_context`. The source that had a custom property with the same name wouldn't be working as expected, right?

File: src/wp-includes/class-wp-block-bindings-registry.php (hunk 16 above)
Comment (SantosGuillamot) [reply src/wp-includes/class-wp-block-bindings-registry.php]: Just did it in [this commit](https://github.com/WordPress/wordpress-develop/pull/6079/commits/59fde2951385ae6af133e44dc334be65a6745ccb).

File: src/wp-includes/class-wp-block-bindings-source.php (hunk 6 above)
Comment (gziolo) [reply src/wp-includes/class-wp-block-bindings-source.php]: I'll keep thinking about it as I don't have an ultimate answer for now as it's hard to tell what will be needed here.

File: src/wp-includes/class-wp-block-bindings-registry.php (hunk 18 above)
Comment (SantosGuillamot) [reply src/wp-includes/class-wp-block-bindings-registry.php:195]: This makes sense to me. Although I don't know how it will work 🤔 I was thinking something like this:

```php
	/**
	 * Source that is being processed.
	 *
	 * @since 6.5.0
	 * @var WP_Block_Bindings_Source|false
	 */
	private $source;

	/**
	 * Callback passed to the `get_block_type_uses_context` filter to add the source's `uses_context` to the block.
	 *
	 * @since 6.5.0
	 *
	 * @param array         $uses_context Array of registered uses context for a block type.
	 * @param WP_Block_Type $block_type   The full block type object.
	 * @return array Modified array of uses_context with the values provided by the source.
	 */
	public function _add_source_uses_context( $uses_context, $block_type ) {
		$source = $this->source;
		if ( ! in_array( $block_type->name, $this->supported_blocks, true ) || empty( $source->uses_context ) ) {
			return $uses_context;
		}
		// Use array_values to reset the array keys.
		return array_values( array_unique( array_merge( $uses_context, $source->uses_context ) ) );
	}
```

But we probably need a different callback name per source.

Anyway, I'll add a task in [the tracking issue](https://github.com/WordPress/gutenberg/issues/53300) to handle it 🙂

**Comment by gziolo** [comment]
Committed with https://core.trac.wordpress.org/changeset/57641.
